### PR TITLE
Resolve clashing atom labels in 2D drawing

### PR DESCRIPTION
**Pull request by DavidACosgrove**
#### Reference Issue
This came up in discussion #8044.

#### What does this implement/fix? Explain your changes.
If 2 atom labels are seen to clash, the orientation of one or the other is changed to try and fix it.
So
image
becomes
image

#### Any other comments?
Probably it could be a bit better, such that the upper methyls went up rather than down but that seemed like a bit more effort than is warranted by the frequency that the issue occurs.

### CODE REVIEW

File: Code/GraphMol/MolDraw2D/AtomSymbol.h [hunk 1]
```diff
@@ -67,6 +67,7 @@ class AtomSymbol {
   // we might be scaling the font differently from the drawing as a whole.
   virtual void scale(const Point2D &scaleFactor);
   virtual void move(const Point2D &trans);
+  virtual void recalculateRects();
```
Comment (greglandrum) [inline Code/GraphMol/MolDraw2D/AtomSymbol.h:70]: does this need to be virtual?

File: Code/GraphMol/MolDraw2D/DrawMol.cpp [hunk 2]
```diff
@@ -953,6 +969,82 @@ void DrawMol::extractCloseContacts() {
...
+
+// ****************************************************************************
+void DrawMol::resolveAtomSymbolClashes() {
+  const static std::vector<OrientType> allOrients{OrientType::N, OrientType::E,
+                                                  OrientType::S, OrientType::W};
+  for (auto at1 : drawMol_->atoms()) {
+    const auto atIdx1 = at1->getIdx();
+    for (auto at2 : drawMol_->atoms()) {
+      const auto atIdx2 = at2->getIdx();
+      if (atIdx1 == atIdx2) {
```
Comment (greglandrum) [inline Code/GraphMol/MolDraw2D/DrawMol.cpp:1012]: Wouldn't this be more efficient (yet still correct) as:
```suggestion
      if (atIdx1 >= atIdx2) {
```
We don't need to compare rect A against rect B and rect B against A, right?

**Review by greglandrum — CHANGES_REQUESTED** [review]
small suggested changes

File: Code/GraphMol/MolDraw2D/DrawMol.cpp (hunk 2 above)
Comment (DavidACosgrove) [reply Code/GraphMol/MolDraw2D/DrawMol.cpp:1012]: I think it’s worse than that. A late “refinement” has it only moving one of the 2 whichever order they’re in. I will re-work it.

File: Code/GraphMol/MolDraw2D/catch_tests.cpp [hunk 3]
```diff
@@ -103,10 +103,10 @@ const std::map<std::string, std::hash_result_t> SVG_HASHES = {
     {"testBrackets-2b.svg", 776446765U},
     {"testBrackets-2c.svg", 1050851904U},
     {"testBrackets-2d.svg", 776446765U},
-    {"testBrackets-3a.svg", 1374944956U},
+    {"testBrackets-3a.svg", 3096284602U},
```
Comment (greglandrum) [inline Code/GraphMol/MolDraw2D/catch_tests.cpp:106]: Did you look at the test drawings that changed to make sure the new drawings make sense?

Comment (DavidACosgrove) [reply Code/GraphMol/MolDraw2D/catch_tests.cpp:106]: I looked at 3a. IIRC the OH had moved.  I don’t recall the other 2. I will check tomorrow morning.

Comment (DavidACosgrove) [reply Code/GraphMol/MolDraw2D/catch_tests.cpp:106]: I think everything is ok.  A couple, such as testHandDrawn-5[ab].svg are improved.

Comment (greglandrum) [reply Code/GraphMol/MolDraw2D/catch_tests.cpp:106]: > I think everything is ok. A couple, such as testHandDrawn-5[ab].svg are improved.

Thanks for checking!

**Pull request merged**